### PR TITLE
Feature/core rewrite

### LINKED ISSUE #85: Enhance documentation about tsconfig.json settings
Mentioned in #84:
We should documentate the used minimum settings for the `tsconfig.json` file.

In addition:
- Create yeoman generator for a full blown app (#81)
- #82

### PR DESCRIPTION
**Pull request by buehler**
Adding dependency injection to the core.
Perform core rewrite to provide more structure in the code

todo:
- [x] Dependency Injection and interfaces
- [x] Move actual code
- [x] Documentation (JSDoc) of interfaces
- [x] Cleanup classes and code structures (registration classes and such)
- [x] Add tests for default implementations
- [x] #91 (since newman tests are failing)

Closes #85

### CODE REVIEW

**Comment by buehler** [comment]
Note: all default implementations are tested, but in different files ;-)
Should we relocate them, or wait until a relocation would be needed?

File: core/DefaultRouteHandler.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+                let handleError = (error: any) => {
+                    if (!(error instanceof Error)) {
+                        error = new GenericRouteError(route.propertyKey, error);
+                    }
+                    errorHandler.handleError(instance, request, response, error);
+                };
+
+                try {
+                    paramValues = this.paramHandler.getParamValuesForRequest(params, request, response);
```
Comment (dwirz) [inline core/DefaultRouteHandler.ts:84]: I know there was a reason to not include this in the try-block below... but what was it again? :)

File: core/DefaultRouteHandler.ts [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                try {
+                    let result = method.apply(instance, paramValues),
+                        responseFunction = result => {
+                            if (NON_JSON_TYPES.indexOf(result.constructor) !== -1) {
+                                response.send(result);
+                            } else {
+                                response.json(result);
+                            }
+                        };
+                    if (!returnType && hasResponseParam) {
```
Comment (dwirz) [inline core/DefaultRouteHandler.ts:94]: I know it's picky, but add some new lines for readability between the ifs..

File: core/DefaultRouteHandler.ts [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            .forEach(r => this.registerRoute(r, router));
+
+        return router;
+    }
+
+    public resetRoutes(): void {
+        this.routes = {};
+    }
+
+    private registerRoute(registration: RegistrationHelper, router: Router) {
```
Comment (dwirz) [inline core/DefaultRouteHandler.ts]: Add `: void` since you added it above.. :)

File: docs/index.md [hunk 4]
```diff
@@ -47,6 +46,37 @@ This package installs the transpiled `*.js` files instead of `*.ts` files, since
 to compile them everytime. The declaration files with the JSDocs are provided aswell so your 
 autocomplete does work as it should.
 
+### tsconfig.json
+
+If you don't use the yeoman generator to scaffold your application, you need
```
Comment (dwirz) [inline docs/index.md]: Add link to giuseppe yeoman generator like below ;)

File: core/DefaultRouteHandler.ts (hunk 1 above)
Comment (buehler) [reply core/DefaultRouteHandler.ts:84]: What exactly is not included in the try block?

**Pull request merged**